### ISSUE TITLE
Redefine channel_max

### ISSUE THREAD

**Issue by carlhoerberg** [pdezwart/php-amqp#144]
Can you redefine channels_max (`#define PHP_AMQP_PROTOCOL_MAX_CHANNELS 256`) in runtime?

**Comment by pinepain** [comment]
What is the use case for that?

**Comment by carlhoerberg** [comment]
Seems like the library can't negotiate down the channel limit of the server has a lower limit.

**Comment by pinepain** [comment]
What AMQP broker do you use? Can you show us server log entries related to you problem and show what kind of exception occurs in php and how it looks like? Can you specify environment and provide us example to reproduce the issue?

As a workaround (if the issue really about max-channels limit) we can add extra parameter to `$credentials` array in `AMQPConnection::__construct()` (which might be good idea even without this issue while it brings extra control over options that are for end-user to userspace).

**Comment by carlhoerberg** [comment]
RabbitMQ

**Comment by pinepain** [comment]
What about logs, env, etc?

**Comment by pinepain** [comment]
@carlhoerberg please provide more details on issue, while I cant reproduce any strange behavior when broker has lower, higher `channel_max` set and when it is default (not defined).

**Comment by tlshaheen** [comment]
@pinepain I'm encountering this problem when accessing RabbitMQ managed by CloudAMQP. Running PHP 5.4 and using the https://github.com/fintech-fab/laravel-queue-rabbitmq interface.

CloudAMQP has `channel_max` restrictions configured for the RabbitMQ servers. So when I try to make a connection to the RabbitMQ server using the default PHP-AMQP/fintech-fab setup, I get rejected because the `channel_max` setting I'm passing is outside the restrictions set by the receiving RabbitMQ server.

CloudAMQP tested this with me by temporarily removing their `channel_max` restrictions for the instance I was attempting to connect to. When they did that, I was able to connect.

If I can configure max-channels at runtime, I can make a connection to any RabbitMQ instance regardless of their max-channels restrictions.

**Comment by pinepain** [comment]
Please, specify which php-amqp and librabbitmq-c do you use.

I configured lower `channel-max` (I tried 128, 2, 1280) on testing RabbitMQ but I didn't experience any issues.

I have "Little Lemur" free plan from CloudAMQP with 3 concurrent connections limit (funny, I can establish 4 of them, didn't tried more) and nothing suspicious.

I realize that manually specifying `channel-max` is pretty good idea but we have to figure out what is the problem exactly, while with RabbitMQ if you specify higher `channel-max` then it will be silently lower (you can check actual `channel-max` by calling `AMQPConnection::getMaxChannels()`.

**Comment by tlshaheen** [comment]
[php-amqp 1.4.0](http://pecl.php.net/package/amqp/1.4.0) and [rabbitmq-c v0.5.2](https://github.com/alanxz/rabbitmq-c/releases/tag/v0.5.2). 

I'm able to make a connection to my Little Lemur free plan as well right now. @carlhoerberg, is the `channel-max` restriction still temporarily removed?

**Comment by carlhoerberg** [comment]
It's temporarily removed yes

**Comment by tlshaheen** [comment]
@carlhoerberg What is the `max-channel` restriction normally set to, and how's it set?

**Comment by pinepain** [comment]
@tlshaheen, please, try [1.6.0beta2](http://pecl.php.net/package/amqp/1.6.0beta2), it has max channels limit set to 256.

I just tried to downgrade to 1.4.0 and that is true that issue exists, this is what i get in RabbitMQ logs:
```
=ERROR REPORT==== 8-Feb-2015::17:54:00 ===
closing AMQP connection <0.1713.0> (127.0.0.1:52922 -> 127.0.0.1:5672):
{handshake_error,tuning,0,
    {exit,
        {amqp_error,not_allowed,
            "negotiated channel_max = 0 is higher than the maximum allowed value (2)",
            'connection.tune'},
```
running simple script

```php
<?php
$connection1 = new AMQPConnection();
$connection1->connect();
$channel1 = new AMQPChannel($connection1);
var_dump($connection1, $channel1);
sleep(100);
```

Again, I cant reproduce this issue on latest [1.6.0beta2](http://pecl.php.net/package/amqp/1.6.0beta2). I highly recommend use latest php-amqp while there are a lot of fixes and refactoring in latest release.

Although, ***this is** an issue. I will work on `channel_max` (as well as `frame_max`) support. ETA is from 1 day to 1 week.

**Comment by tlshaheen** [comment]
@pinepain Thanks for looking into this. I've upgraded to [1.6.0beta2](http://pecl.php.net/package/amqp/1.6.0beta2) and it's working. I look forward to seeing the channel_max support!

**Comment by pinepain** [comment]
Update on issue: I've added `frame_max` and `heartbeat` values gettters to rabbitmq-c (https://github.com/alanxz/rabbitmq-c/pull/244) and when new version of rabbitmq-c will be released, I will add control over `channel_max`, `frame_max` and `heartbeat` (as well as partial support for heartbeats - partial because of how rabbitmq-c handles it - only during blocking calls).

I guess ETA can be updated to new value, say, from 2 days to 2 weeks as of now (so deadline is no later than Mar 1,  2015 :smile: ).

**Comment by stof** [comment]
@pinepain I tried installing the beta version of 1.6 on my machine, but I get an error:

```
PHP Warning:  PHP Startup: Unable to load dynamic library '/usr/lib/php5/20131226/amqp.so' -
/usr/lib/php5/20131226/amqp.so: undefined symbol: amqp_basic_nack in Unknown on line 0
```

Is php-amqp 1.6 incompatible with librabbitmq 0.4.1 (which is the one available on Ubuntu 14.04) ?

**Comment by lstrojny** [comment]
@stof yep, at least librabbitmq 0.5.2 is now required
